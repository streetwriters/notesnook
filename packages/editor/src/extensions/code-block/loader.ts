/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { loadLanguage as _loadLanguage } from "./languages/index.js";

const loadedLanguages: Record<string, boolean> = {};
export function isLanguageLoaded(name: string) {
  return !!loadedLanguages[name];
}
export async function loadLanguage(shortName: string) {
  const { default: language } = (await _loadLanguage(shortName)) || {};
  loadedLanguages[shortName] = true;
  return language;
}
