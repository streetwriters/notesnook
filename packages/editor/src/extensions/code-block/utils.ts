/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Node } from "@tiptap/pm/model";
import { Selection } from "@tiptap/pm/state";
import type { CodeBlockAttributes } from "./code-block.js";

export type CodeLine = {
  index: number;
  from: number;
  to: number;
  length: number;
  text: (length?: number) => string;
};

export type CaretPosition = {
  column: number;
  line: number;
  selected?: number;
  total: number;
  from: number;
};

export function toCodeLines(code: string, pos: number): CodeLine[] {
  const positions: CodeLine[] = [];

  let start = 0;
  let from = pos + 1;
  let index = 0;
  while (start <= code.length) {
    let end = code.indexOf("\n", start);
    if (end <= -1) end = code.length;

    const lineLength = end - start;
    const to = from + lineLength;
    const lineStart = start;
    positions.push({
      index,
      length: lineLength,
      from,
      to,
      text: (length) => {
        return code.slice(
          lineStart,
          length ? lineStart + length : lineStart + lineLength
        );
      }
    });

    from = to + 1;
    start = end + 1;
    ++index;
  }
  return positions;
}

export function toCaretPosition(
  selection: Selection,
  lines?: CodeLine[]
): CaretPosition | undefined {
  const { $from, $to, $head } = selection;
  if ($from.parent.type.name !== "codeblock") return;
  lines = lines || getLines($from.parent);

  for (const line of lines) {
    if ($head.pos >= line.from && $head.pos <= line.to) {
      const lineLength = line.length + 1;
      return {
        line: line.index + 1,
        column: lineLength - (line.to - $head.pos),
        selected: $to.pos - $from.pos,
        total: lines.length,
        from: line.from
      };
    }
  }
  return;
}

export function getLines(node: Node) {
  const { lines } = node.attrs as CodeBlockAttributes;
  return lines || [];
}
