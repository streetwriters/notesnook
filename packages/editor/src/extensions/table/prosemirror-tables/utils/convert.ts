import type { Node } from "prosemirror-model";
import { TableMap } from "../tablemap.js";

/**
 * This function will transform the table node into a matrix of rows and columns
 * respecting merged cells, for example this table:
 *
 * ```
 * ┌──────┬──────┬─────────────┐
 * │  A1  │  B1  │     C1      │
 * ├──────┼──────┴──────┬──────┤
 * │  A2  │     B2      │      │
 * ├──────┼─────────────┤  D1  │
 * │  A3  │  B3  │  C3  │      │
 * └──────┴──────┴──────┴──────┘
 * ```
 *
 * will be converted to the below:
 *
 * ```javascript
 * [
 *   [A1, B1, C1, null],
 *   [A2, B2, null, D1],
 *   [A3, B3, C3, null],
 * ]
 * ```
 * @internal
 */
export function convertTableNodeToArrayOfRows(
  tableNode: Node
): (Node | null)[][] {
  const map = TableMap.get(tableNode);
  const rows: (Node | null)[][] = [];
  const rowCount = map.height;
  const colCount = map.width;
  for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
    const row: (Node | null)[] = [];
    for (let colIndex = 0; colIndex < colCount; colIndex++) {
      const cellIndex = rowIndex * colCount + colIndex;
      const cellPos = map.map[cellIndex];
      if (rowIndex > 0) {
        const topCellIndex = cellIndex - colCount;
        const topCellPos = map.map[topCellIndex];
        if (cellPos === topCellPos) {
          row.push(null);
          continue;
        }
      }
      if (colIndex > 0) {
        const leftCellIndex = cellIndex - 1;
        const leftCellPos = map.map[leftCellIndex];
        if (cellPos === leftCellPos) {
          row.push(null);
          continue;
        }
      }
      row.push(tableNode.nodeAt(cellPos));
    }
    rows.push(row);
  }

  return rows;
}

/**
 * Convert an array of rows to a table node.
 *
 * @internal
 */
export function convertArrayOfRowsToTableNode(
  tableNode: Node,
  arrayOfNodes: (Node | null)[][]
): Node {
  const newRows: Node[] = [];
  const map = TableMap.get(tableNode);
  const rowCount = map.height;
  const colCount = map.width;
  for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
    const oldRow: Node = tableNode.child(rowIndex);
    const newCells: Node[] = [];

    for (let colIndex = 0; colIndex < colCount; colIndex++) {
      const cell = arrayOfNodes[rowIndex][colIndex];
      if (!cell) {
        continue;
      }

      const cellPos = map.map[rowIndex * map.width + colIndex];
      const oldCell = tableNode.nodeAt(cellPos);
      if (!oldCell) {
        continue;
      }

      const newCell = oldCell.type.createChecked(
        cell.attrs,
        cell.content,
        cell.marks
      );
      newCells.push(newCell);
    }

    const newRow = oldRow.type.createChecked(
      oldRow.attrs,
      newCells,
      oldRow.marks
    );
    newRows.push(newRow);
  }

  const newTable = tableNode.type.createChecked(
    tableNode.attrs,
    newRows,
    tableNode.marks
  );
  return newTable;
}
