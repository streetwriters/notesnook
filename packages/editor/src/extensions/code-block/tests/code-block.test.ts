/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { createEditor, h } from "../../../../test-utils";
import { expect, test, vi } from "vitest";
import { CodeBlock, inferLanguage } from "../code-block";
import { HighlighterPlugin } from "../highlighter";
import { getChangedNodes } from "../../../utils/prosemirror";
import { refractor } from "refractor/lib/core";

const CODEBLOCKS_HTML = h("div", [
  h("pre", [h("code", ["function hello() { }"])], {
    class: "language-javascript"
  }),
  h("pre", [h("code", ["function hello() { }"])], {
    class: "language-typescript"
  })
]).innerHTML;

const CODE_PARAGRAPH = `function helloWorld() {
  console.log('Hello, World!');
}`;

interface DocumentNode {
  type: string;
  attrs?: { [key: string]: any };
  content?: DocumentNode[];
  text?: string;
}

test("codeblocks should get highlighted on init", async () => {
  const editorElement = h("div");
  createEditor({
    element: editorElement,
    initialContent: CODEBLOCKS_HTML,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();
});

test("codeblocks should get highlighted after pasting", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  editor.commands.setContent(CODEBLOCKS_HTML, true, {
    preserveWhitespace: true
  });

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();
});

test("codeblocks should not be updated if other content is changed", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    initialContent: CODEBLOCKS_HTML,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  await new Promise((resolve) => setTimeout(resolve, 100));

  const onTransactionMock = vi.fn();
  editor.on("transaction", ({ transaction }) => {
    const changedNodes = getChangedNodes(transaction).map(
      (n) => n.node.type.name
    );
    onTransactionMock(changedNodes);
  });

  editor.commands.insertContentAt(
    editor.state.doc.nodeSize - 2,
    h("p", ["A new paragraph."]).outerHTML
  );

  expect(onTransactionMock).toHaveBeenCalledOnce();
  expect(onTransactionMock).toHaveBeenCalledWith(["paragraph"]);
  expect(editorElement.outerHTML).toMatchSnapshot();
});

test("pasting code from vscode should automatically create a syntax highlighted codeblock", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  const clipboardEvent = new Event("paste", {
    bubbles: true,
    cancelable: true,
    composed: true
  });

  (clipboardEvent as unknown as any)["clipboardData"] = {
    getData: (type: string) =>
      type === "text/plain"
        ? `function hello()
{
  const world = "hello";
}`
        : type === "vscode-editor-data"
        ? JSON.stringify({ mode: "javascript" })
        : undefined
  };

  editor.view.dom.dispatchEvent(clipboardEvent);

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();
});

test("infer code language from class", async () => {
  expect(
    inferLanguage(h("pre", [], { class: "language-js hello world hope" }))
  ).toBe("javascript");

  expect(
    inferLanguage(h("pre", [], { class: "lang-typescript hello world hope" }))
  ).toBe("typescript");

  expect(
    inferLanguage(h("pre", [], { class: "brush:js hello world hope" }))
  ).toBe("javascript");

  expect(
    inferLanguage(h("pre", [], { class: "brush: js hello world hope" }))
  ).toBe("javascript");

  expect(inferLanguage(h("pre", [], { lang: "javascript" }))).toBe(
    "javascript"
  );

  expect(
    inferLanguage(
      h("pre", [], { class: "brush: js hello world hope", lang: "typescript" })
    )
  ).toBe("javascript");
});

test("editing code in a highlighted code block should not be too slow", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    initialContent: CODEBLOCKS_HTML,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  const plugin = HighlighterPlugin({
    name: CodeBlock.name,
    defaultLanguage: "plaintext"
  });
  let highlighterState = plugin.spec.state?.init(editor.state, editor.state);

  expect(highlighterState?.decorations).toBeDefined();
  expect(highlighterState?.languages).toBeDefined();

  const timings: number[] = [];
  for (let i = 0; i < 100; ++i) {
    const { tr } = editor.state;
    tr.insertText(`const hello${i} = ${i};`, tr.doc.nodeSize - 3);
    const now = Date.now();
    highlighterState = plugin.spec.state?.apply(
      tr,
      highlighterState!,
      editor.state,
      editor.state
    );
    editor.view.dispatch(tr);
    timings.push(Date.now() - now);
  }

  expect(timings.reduce((a, b) => a + b) / timings.length).toBeLessThan(16);
  expect(editorElement.outerHTML).toMatchSnapshot();
});

test("Adding a new codeblock & changing the language should apply the new highlighting", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  editor.commands.setCodeBlock();
  editor.commands.insertContent("function hello() { }");

  editor.commands.updateAttributes(CodeBlock.name, { language: "javascript" });

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();
  expect(refractor.registered("javascript")).toBe(true);
});

test("Switching codeblock language should register the new language", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    initialContent: CODEBLOCKS_HTML,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });
  editor.commands.updateAttributes(CodeBlock.name, { language: "java" });
  await new Promise((resolve) => setTimeout(resolve, 100));
  expect(refractor.registered("java")).toBe(true);
});

test("Toggling multiple paragraphs into a codeblock should make one codeblock", async () => {
  const editorElement = h("div");
  const { editor } = createEditor({
    element: editorElement,
    initialContent: CODE_PARAGRAPH,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  const docLength = editor.state.doc.nodeSize - 2;

  editor.commands.setTextSelection({ from: 0, to: docLength });
  editor.commands.toggleCodeBlock();
  editor.commands.updateAttributes(CodeBlock.name, { language: "javascript" });

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();

  const docJSON = editor.state.doc.toJSON();
  const content = docJSON.content;
  const codeBlockNodes = content.filter(
    (node: DocumentNode) => node.type === "codeblock"
  );
  const otherNodes = content.filter(
    (node: DocumentNode) => node.type !== "codeblock"
  );

  expect(codeBlockNodes).toHaveLength(1);
  expect(otherNodes).toHaveLength(0);
});

test("Toggling from a codeblock should make codeblock disappear", async () => {
  const editorElement = document.createElement("div");
  const { editor } = createEditor({
    element: editorElement,
    initialContent: CODEBLOCKS_HTML,
    extensions: {
      codeblock: CodeBlock,
      codeBlock: false,
      code: false
    }
  });

  const docLength = editor.state.doc.nodeSize - 2;

  editor.commands.setTextSelection({ from: 0, to: docLength });

  editor.commands.toggleCodeBlock();

  await new Promise((resolve) => setTimeout(resolve, 100));

  expect(editorElement.outerHTML).toMatchSnapshot();

  const docJSON = editor.state.doc.toJSON();
  const content = docJSON.content;
  const codeBlockNodes = content.filter(
    (node: DocumentNode) => node.type === "codeblock"
  );
  const paragraphNodes = content.filter(
    (node: DocumentNode) => node.type === "paragraph"
  );

  expect(codeBlockNodes).toHaveLength(0);
  expect(paragraphNodes).toHaveLength(2);
});
