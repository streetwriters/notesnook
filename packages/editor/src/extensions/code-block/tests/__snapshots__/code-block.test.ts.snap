// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Adding a new codeblock & changing the language should apply the new highlighting 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-xyz" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div></div></div>"`;

exports[`Toggling from a codeblock should make codeblock disappear 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></p><p><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></p></div></div>"`;

exports[`Toggling multiple paragraphs into a codeblock should make one codeblock 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-xyz" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">helloWorld</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> console<span class="token punctuation">.</span><span class="token function">log</span><span class="token punctuation">(</span><span class="token string">'Hello, World!'</span><span class="token punctuation">)</span><span class="token punctuation">;</span> <span class="token punctuation">}</span></pre></div></div></div>"`;

exports[`codeblocks should get highlighted after pasting 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-javascript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-typescript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div></div></div>"`;

exports[`codeblocks should get highlighted on init 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-javascript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-typescript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div></div></div>"`;

exports[`codeblocks should not be updated if other content is changed 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-javascript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-typescript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div><p>A new paragraph.</p></div></div>"`;

exports[`editing code in a highlighted code block should not be too slow 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-javascript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span></pre></div><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-typescript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span> <span class="token punctuation">}</span><span class="token keyword">const</span> hello0 <span class="token operator">=</span> <span class="token number">0</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello1 <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello2 <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello3 <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello4 <span class="token operator">=</span> <span class="token number">4</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello5 <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello6 <span class="token operator">=</span> <span class="token number">6</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello7 <span class="token operator">=</span> <span class="token number">7</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello8 <span class="token operator">=</span> <span class="token number">8</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello9 <span class="token operator">=</span> <span class="token number">9</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello10 <span class="token operator">=</span> <span class="token number">10</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello11 <span class="token operator">=</span> <span class="token number">11</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello12 <span class="token operator">=</span> <span class="token number">12</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello13 <span class="token operator">=</span> <span class="token number">13</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello14 <span class="token operator">=</span> <span class="token number">14</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello15 <span class="token operator">=</span> <span class="token number">15</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello16 <span class="token operator">=</span> <span class="token number">16</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello17 <span class="token operator">=</span> <span class="token number">17</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello18 <span class="token operator">=</span> <span class="token number">18</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello19 <span class="token operator">=</span> <span class="token number">19</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello20 <span class="token operator">=</span> <span class="token number">20</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello21 <span class="token operator">=</span> <span class="token number">21</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello22 <span class="token operator">=</span> <span class="token number">22</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello23 <span class="token operator">=</span> <span class="token number">23</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello24 <span class="token operator">=</span> <span class="token number">24</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello25 <span class="token operator">=</span> <span class="token number">25</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello26 <span class="token operator">=</span> <span class="token number">26</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello27 <span class="token operator">=</span> <span class="token number">27</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello28 <span class="token operator">=</span> <span class="token number">28</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello29 <span class="token operator">=</span> <span class="token number">29</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello30 <span class="token operator">=</span> <span class="token number">30</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello31 <span class="token operator">=</span> <span class="token number">31</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello32 <span class="token operator">=</span> <span class="token number">32</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello33 <span class="token operator">=</span> <span class="token number">33</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello34 <span class="token operator">=</span> <span class="token number">34</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello35 <span class="token operator">=</span> <span class="token number">35</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello36 <span class="token operator">=</span> <span class="token number">36</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello37 <span class="token operator">=</span> <span class="token number">37</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello38 <span class="token operator">=</span> <span class="token number">38</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello39 <span class="token operator">=</span> <span class="token number">39</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello40 <span class="token operator">=</span> <span class="token number">40</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello41 <span class="token operator">=</span> <span class="token number">41</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello42 <span class="token operator">=</span> <span class="token number">42</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello43 <span class="token operator">=</span> <span class="token number">43</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello44 <span class="token operator">=</span> <span class="token number">44</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello45 <span class="token operator">=</span> <span class="token number">45</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello46 <span class="token operator">=</span> <span class="token number">46</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello47 <span class="token operator">=</span> <span class="token number">47</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello48 <span class="token operator">=</span> <span class="token number">48</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello49 <span class="token operator">=</span> <span class="token number">49</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello50 <span class="token operator">=</span> <span class="token number">50</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello51 <span class="token operator">=</span> <span class="token number">51</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello52 <span class="token operator">=</span> <span class="token number">52</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello53 <span class="token operator">=</span> <span class="token number">53</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello54 <span class="token operator">=</span> <span class="token number">54</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello55 <span class="token operator">=</span> <span class="token number">55</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello56 <span class="token operator">=</span> <span class="token number">56</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello57 <span class="token operator">=</span> <span class="token number">57</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello58 <span class="token operator">=</span> <span class="token number">58</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello59 <span class="token operator">=</span> <span class="token number">59</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello60 <span class="token operator">=</span> <span class="token number">60</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello61 <span class="token operator">=</span> <span class="token number">61</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello62 <span class="token operator">=</span> <span class="token number">62</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello63 <span class="token operator">=</span> <span class="token number">63</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello64 <span class="token operator">=</span> <span class="token number">64</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello65 <span class="token operator">=</span> <span class="token number">65</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello66 <span class="token operator">=</span> <span class="token number">66</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello67 <span class="token operator">=</span> <span class="token number">67</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello68 <span class="token operator">=</span> <span class="token number">68</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello69 <span class="token operator">=</span> <span class="token number">69</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello70 <span class="token operator">=</span> <span class="token number">70</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello71 <span class="token operator">=</span> <span class="token number">71</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello72 <span class="token operator">=</span> <span class="token number">72</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello73 <span class="token operator">=</span> <span class="token number">73</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello74 <span class="token operator">=</span> <span class="token number">74</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello75 <span class="token operator">=</span> <span class="token number">75</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello76 <span class="token operator">=</span> <span class="token number">76</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello77 <span class="token operator">=</span> <span class="token number">77</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello78 <span class="token operator">=</span> <span class="token number">78</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello79 <span class="token operator">=</span> <span class="token number">79</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello80 <span class="token operator">=</span> <span class="token number">80</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello81 <span class="token operator">=</span> <span class="token number">81</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello82 <span class="token operator">=</span> <span class="token number">82</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello83 <span class="token operator">=</span> <span class="token number">83</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello84 <span class="token operator">=</span> <span class="token number">84</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello85 <span class="token operator">=</span> <span class="token number">85</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello86 <span class="token operator">=</span> <span class="token number">86</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello87 <span class="token operator">=</span> <span class="token number">87</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello88 <span class="token operator">=</span> <span class="token number">88</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello89 <span class="token operator">=</span> <span class="token number">89</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello90 <span class="token operator">=</span> <span class="token number">90</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello91 <span class="token operator">=</span> <span class="token number">91</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello92 <span class="token operator">=</span> <span class="token number">92</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello93 <span class="token operator">=</span> <span class="token number">93</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello94 <span class="token operator">=</span> <span class="token number">94</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello95 <span class="token operator">=</span> <span class="token number">95</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello96 <span class="token operator">=</span> <span class="token number">96</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello97 <span class="token operator">=</span> <span class="token number">97</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello98 <span class="token operator">=</span> <span class="token number">98</span><span class="token punctuation">;</span><span class="token keyword">const</span> hello99 <span class="token operator">=</span> <span class="token number">99</span><span class="token punctuation">;</span></pre></div></div></div>"`;

exports[`pasting code from vscode should automatically create a syntax highlighted codeblock 1`] = `
"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><div class="codeblock-view-content-wrap"><pre class="node-content-wrapper language-javascript" style="white-space: pre; min-width: 20px;"><span class="token keyword">function</span> <span class="token function">hello</span><span class="token punctuation">(</span><span class="token punctuation">)</span>
<span class="token punctuation">{</span>
  <span class="token keyword">const</span> world <span class="token operator">=</span> <span class="token string">"hello"</span><span class="token punctuation">;</span>
<span class="token punctuation">}</span></pre></div></div></div>"
`;
